edb/server/protocol/system_api.py

```diff
@@ -0,0 +1,153 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2019-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import http
+import json
+
+import immutables as immu
+
+from edb import errors
+from edb import edgeql
+
+from edb.common import debug
+from edb.common import markup
+
+from edb.schema import schema as s_schema
+
+from edb.server import compiler
+from edb.server.compiler import IoFormat
+from edb.server.compiler import enums
+from edb.server import defines as edbdef
+
+
+ALLOWED_CAPABILITIES = (
+    enums.Capability.MODIFICATIONS
+)
+
+
+async def handle_request(
+    request,
+    response,
+    path_parts,
+    server,
+):
+    try:
+        if path_parts == ['status', 'ready'] and request.method == b'GET':
+            await handle_status_request(request, response, server)
+        else:
+            response.body = b'Unknown path'
+            response.status = http.HTTPStatus.NOT_FOUND
+            response.close_connection = True
+
+        return
+    except Exception as ex:
+        if debug.flags.server:
+            markup.dump(ex)
+
+        ex_type = type(ex)
+        if not issubclass(ex_type, errors.EdgeDBError):
+            # XXX Fix this when LSP "location" objects are implemented
+            ex_type = errors.InternalServerError
+
+        err_dct = {
+            'message': str(ex),
+            'type': str(ex_type.__name__),
+            'code': ex_type.get_code(),
+        }
+
+        response.body = json.dumps({'error': err_dct}).encode()
+        response.status = http.HTTPStatus.INTERNAL_SERVER_ERROR
+        response.close_connection = True
+        return
+
+
+async def handle_status_request(
+    request,
+    response,
+    server,
+):
+    result = await execute(server, "SELECT 'OK'", {})
```

Comment:
An unhandled exception would propagate and cause the server to return a 400, text/plain response. Is that acceptable? Maybe it should always return application/json?

Reply:
No, errors are handled in the caller, see above.

Reply:
Ah, I forgot my own spaghetti. You're right.